planning/MartenV4-WastazComments.md

```diff
@@ -0,0 +1,174 @@
+# Projections
+
+## Building projections
+While its nice that Marten provides several options for actually constructing projections (`ITransform`, `AggregateStreamWith`, `ViewProjection`, 
+raw `IProjection`, etc...honestly Ive probably forgotten some way of doing it) in the end the ones that I end up using 99% of the time is
+either `ViewProjection` or implementing the raw `IProjection`. The main reason for this is that most of the other options are too limiting
+when building projections. 
+
+Working with `ViewProjection` is quite nice, however occasionally the number of overloads can make it hard because its absolutely not hard
+to make the compiler confused over which overload you are actually trying to use. The main thing that I cant do with `ViewProjection` which
+often is the thing that makes me drop down to `IProjection` is project an event to multiple documents, which is something that can be very useful
+for solving some problems. 
+
+Working with `IProjection` often feels very raw, and I have to admit that I still havent figured out the actual difference between `Apply` and `ApplyAsync`
+because it seems like they are...executed depending on if the projection runs inline or via the async daemon? Thats just confusing tbh.
+
+Also (and now we are getting ahead of myself), projections depend on mutating state. I think this is fine, even fine as a default. But I do wish that
+there would be an option to return an entire new object instead of mutating. This would also help a lot in the F# world.
+
+## Running projections
+Its great that Marten supports both running the projections inline and async. While Im suspecting that some really hardcore cqrs/es people
+balk at the prospect of inline projections, they are a great option to have in some cases even though I try hard not to overuse them.
+
+The async daemon however, is a bit lacking (though it does its job ok). The documentation is an old blog post by Jeremy which doesnt say
+that much about it tbh. And you cant shard or put up multiples of it for some sort of HA scenario. It feels a bit scary just running a single service
+with the daemon in production per database. The error handling is a bit funky as well, and if you have an even that crashes a projection somewhere
+in the middle of an event page then actually singling out _that single event_ to have a look at it properly can be very frustrating. I have so far been
+lucky enough to find these weird events in my QA environment so I havent had to do that in production _yet_ but I do not look forward to doing that.
+
+Replaying projections seems to be a thing that you can do, but I find the documentation lacking on this so Im a bit unsure of what actually this does
+which tends to make me prefer spinning up a new projection in parallel and then killing off the old one when it is in sync. It feels safer. But maybe it isnt?
+
+## Unit testing projections
```

Comment:
100 that! As I stated in other places - this is must have to make it better testable (or in general testable without the Marten infrastructure). Most of the other ES frameworks allow that - it's kinda standard.